### PR TITLE
Add deprecated keyword to deprecate any symbol

### LINKED ISSUE #14396: Design alternative to function overloads for deprecating non-function symbols
Context: I tried to add a deprecation warning for enums defined in the IO module in such a way that it only fired when the module was not explicitly used, and would not fire when an explicit use of IO was available to that scope.  I tried this by adding an argumentless function with the same name as the enum, which would warn and then return the IO module's type.  This kinda worked, but there were a few issues:
- We couldn't provide the same warning for the enum's constants, so it wasn't useful to warn because the program would just fail anyways
- While we always got the error when IO was not used, and never got the error when IO was used in an unlimited manner, `use IO only iomode;` did not allow us to access the enum and instead triggered the error message.  Similarly `IO.iomode.r` also triggered the error message.

Now, there's definitely some issues there about allowing functions to share the same name as types (and potentially variables, but that's another discussion), and potentially about how limitation clauses on use statements should respond even if we continue to allow that name sharing.  But given the other issue about not being able to warn for the enum's constants, and given that using a function in this case is kinda already a hack, maybe another solution could be used?

Brad suggested we extend "last resort" to apply to modules, and then use modules to provide deprecation warnings for things that functions can't support due to symbol conflicts.  This seems potentially tractable to me

### PR DESCRIPTION
**Pull request by lydia-duncan**
We were previously able to deprecate a proc or iter by using a
compilerWarning within the body.  This worked well, but
deprecating other symbols was more complicated and in some
cases impossible.  This change adds new syntax to deprecate
a symbol, which can be applied to any symbol in a uniform way.
As an example,

```chapel
deprecated
module M { ... }
```

will use a default message of "M is deprecated" when M is
referenced.  An explicit message can be specified by following
the new keyword with a string message, e.g.

```chapel
deprecated "M is deprecated, please use N instead"
module M  { ... }
```

Resolves Cray/chapel-private#2008
Resolves #14396 

<details>

Adds a new flag, FLAG_DEPRECATED, and a new field to the Symbol
base class, as well as a method on Symbol that will return either the
specified message or a computed default message.  The deprecation
message value is copied over into the management class wrapper and
generic instantiations of a function.

Add productions to the parser to support the keyword and message
forms on module level symbols, class level symbols, and enum constants.
These forms allow the symbols to be built as normal, then attach the
deprecation message and flag to the symbol afterwards.

The remaining work is adding checks to the flag in all the places where a
symbol is resolved - when determining the symbol being referenced in an
import or use statement, when validating the only/except lists or import
list for unqualified access, when resolving enum qualified references, and
at function resolution.  The latter involves excluding the generation of the
deprecation warning when we find a use within a compiler generated
function (like `init=`), which was triggering some additional errors.

While here, updated the compiler/next parser to also reflect the changes
(though not fully implement them yet)

Adds a number of tests of this new feature to ensure it behaves
appropriately, including:
- with classes
- with const variable declarations
- that we don't allow deprecation warnings on forwarding statements,
which is the only non-symbol syntax that the parser would allow currently
- that a deprecated outer module will generate a warning if listed as part
of the path of an inner module
- with an enum declaration
- with an enum's constant
- with a field
- with an initializer
- with a method
- with a module
- with an operator
- with a param variable declaration
- with a proc
- with a record
- with a type alias declaration
- with a variable declaration
- with a variable declaration without specifying the message
- with an enum constant without specifying the message
- with a field without specifying the message (both with and without an explicit
initializer)
- various combinations of use in an import and use statement's limitation clause,
including when renaming the symbol
- and the ordering of the keyword relative to pragmas

</details>

Running a full paratest with futures now

### CODE REVIEW

File: test/deprecated/functionality/README [hunk 1]
```diff
@@ -0,0 +1,3 @@
+This directory is for tests of how to deprecate a symbol.  Its contents are
```
Comment (mppf) [inline test/deprecated/functionality/README:1]: Hmm. I'd maybe put this directory in `test/parsing/deprecated` instead or something like that.

**Review by mppf — APPROVED** [review]
Looks good. I'm not sure if it makes sense to update `compiler/next/lib/frontend/Parser/chapel.ypp` in this PR but I suspect it'd at least worth adding some comments to that file. The structure of the parser there should be very similar and you should be able to easily find places you modified and a TODO comment to bring in the deprecation syntax.

File: test/deprecated/functionality/README (hunk 1 above)
Comment (lydia-duncan) [reply test/deprecated/functionality/README:1]: What about a new directory `test/deprecated-keyword` as a sibling of `test/deprecated`?

Comment (mppf) [reply test/deprecated/functionality/README:1]: That would be OK, modulo the issue of it making tab-completion more annoying. I recommend you ask 1 other person; if they like what you originally did better then maybe we stick with that.

Comment (lydia-duncan) [reply test/deprecated/functionality/README:1]: I asked David L and he agrees `test/deprecated/functionality` is bad and likes `test/deprecated-keyword` - autocomplete will stop at `test/deprecated`, which is the one people are more likely to want

**Comment by lydia-duncan** [comment]
I've updated the compiler/next parser and it passes the tests.  I do get:

```
chapel.ypp: warning: 1 nonterminal useless in grammar [-Wother]
chapel.ypp: warning: 3 rules useless in grammar [-Wother]
chapel.ypp:769.1-20: warning: nonterminal useless in grammar: deprecated_decl_stmt [-Wother]
  769 | deprecated_decl_stmt:
```
when running `make libchplcomp-parser` but it seems to not have negatively impacted anything so whatever?

I also get a bunch of path capitalization warnings when running `make libchplcomp` but again, `make test-libchplcomp` seemed to work just fine despite them.  I'm happy to help out with the eventual implementation of the build stage for deprecated symbols later

**Pull request merged**